test/hotspot/jtreg/compiler/vectorapi/reshape/tests/TestVectorCast.java

```diff
@@ -0,0 +1,1359 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.vectorapi.reshape.tests;
+
+import compiler.lib.ir_framework.IR;
+import compiler.lib.ir_framework.Run;
+import compiler.lib.ir_framework.Test;
+
+import static compiler.vectorapi.reshape.utils.VectorReshapeHelper.*;
+import static jdk.incubator.vector.VectorOperators.*;
+
+/**
+ * This class contains all possible cast operations between different vector species.
+ * The methods only take into consideration the actual cast in C2, as the vectors are
+ * ofter shrunk or expanded before/after casting if the element numbers mismatch.
+ * In each cast, the VectorCastNode is expected to appear exactly once.
+ */
+public class TestVectorCast {
```

Comment:
At some point i would like to explore autogenerating such files., from another Java program and a template mechanism (unfortunately we cannot use a library like Java Poet). Not now though! 

This also relates to redesigning the existing vector unit tests, moving away from a bash script to a more flexible Java program, from which we can generate unit tests and/or IR tests.

The challenge with the IR tests is knowing associated IR nodes and the set supported on various platforms. I like how you enumerated the list for the conversions. Ideally we could go to the source of truth in C2 and determine that, but it does not seem easy to determine.

Reply:
I believe if we have a way to query for compiler support from Java the same way as we do in C2 then the task would be trivial. I don't know if we have a way to call into C2 from Java similar to JNI, though.

Reply:
Perhaps a good way to explore that is through the `WhiteBox` API?